bench.py

```diff
@@ -0,0 +1,416 @@
+    if (len(prossessedFiles) == 0):
+        print("No files were processed. Consider changing the filter or nGPU command line arguments.")
+    else:
+        print(f"Processed Files = {prossessedFiles}")
+
+
+# Execute the test program on the file filename, place results in the output folder
+# then returns a list of the names of the output files
+def run_benchmark(mode, algorithmAttributes, filename, endBufSize, inputDirectory, outputDirectory) -> list:
+    print(f"\nRun Benchmark in mode {mode} for file: " + filename)
+
+    # The list of output files that will be returned (one for NCCl and one for MSCCL)
+    outputFileList = []
+
+    # Extract the substring name of the algorithm up until the first "." in the file name. Call this substring "algorithm"
+    mscclCollective = algorithmAttributes.get("coll")
+
+    # Construct the string name of the nccl algorithm from the msccl algorithm.
+    # Nccl uses "_" to separate words in the algorithm name and msccl names are more insconsistent.
+    # For example, msccl uses "all_reduce" and nccl uses "Allreduce".
+    # This code has explicit conversions and will raise an exception if it doesn't know how to convert.
+    # This code uses the python 3.10 match statement if the python version is 3.10 or greater.
+    ncclAlgorithm = ""
+    #    match mscclAlgorithm.lower():
+    #        case "allreduce":
+    #            ncclAlgorithm = "all_reduce"
+    #        case "allgather":
+    #            ncclAlgorithm = "all_gather"
+    #        case "alltoall":
+    #            ncclAlgorithm = "all_to_all"
+    #        case _:
+    #            raise Exception("Unknown algorithm: " + mscclAlgorithm)
+    if mscclCollective.lower() == "allreduce":
+        ncclAlgorithm = "all_reduce"
+    elif mscclCollective.lower() == "allgather":
+        ncclAlgorithm = "all_gather"
+    elif mscclCollective.lower() == "alltoall":
+        ncclAlgorithm = "all_to_all"
+    else:
+        raise Exception("Unexpected algorithm: " + mscclAlgorithmNamePrefix)
+
+    ncclPerfTest = ncclAlgorithm + "_perf"
+
+    # Construct the command to run the benchmarks using the mpirun command
+    # example command:
+    #    mpirun --bind-to numa --tag-output --allow-run-as-root -np 8  -mca pml ob1 -mca btl ^openib -mca btl_tcp_if_include eth0 -x PATH -x LD_LIBRARY_PATH=~/msccl/build/lib/:$LD_LIBRARY_PATH -x UCX_IB_ENABLE_CUDA_AFFINITY=n -x NCCL_IB_PCI_RELAXED_ORDERING=1 -x UCX_IB_PCI_RELAXED_ORDERING=on -x UCX_NET_DEVICES=mlx5_0:1 -x UCX_TLS=rc -x NCCL_SOCKET_IFNAME=eth0 -x NCCL_DEBUG=INFO -x NCCL_NET_GDR_LEVEL=5 -x NCCL_DEBUG_SUBSYS=INIT,ENV -x NCCL_ALGO=MSCCL,RING,TREE -x CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 -x MSCCL_XML_FILES=/home/saemal/test.xml ~/nccl-tests/build/all_reduce_perf -b 1KB -e 32MB -f 2 -g 1 -c 1 -w 100 -n 100
+    #
+    # below are important components of the command
+    #
+    commandPrefix = "mpirun --bind-to numa --tag-output --allow-run-as-root"
```

Comment:
can we take mpirun location as an input?

Reply:
Yes, I've done this as we've discussed.